extensions/sinks/kafka/kafka.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2023 carlclone@gmail.com
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/lf-edge/ekuiper/pkg/api"
+	"github.com/lf-edge/ekuiper/pkg/cast"
+	kafkago "github.com/segmentio/kafka-go"
+	"github.com/segmentio/kafka-go/sasl"
+	"github.com/segmentio/kafka-go/sasl/plain"
+	"github.com/segmentio/kafka-go/sasl/scram"
+	"strings"
+	"time"
+)
+
+type kafkaSink struct {
+	writer *kafkago.Writer
+	c      *sinkConf
+}
+
+const (
+	AT_MOST_ONCE  = "AT_MOST_ONCE"
+	AT_LEAST_ONCE = "AT_LEAST_ONCE"
+
+	SASL_NONE  = "none"
+	SASL_PLAIN = "plain"
+	SASL_SCRAM = "scram"
+)
+
+type sinkConf struct {
+	Brokers           string `json:"brokers"`
+	Topic             string `json:"topic"`
+	DeliveryGuarantee string `json:"deliveryGuarantee"`
+	SaslAuthType      string `json:"saslAuthType"`
+	SaslUserName      string `json:"saslUserName"`
+	SaslPassword      string `json:"saslPassword"`
+}
+
+func (m *kafkaSink) Configure(props map[string]interface{}) error {
+	c := &sinkConf{
+		Brokers:           "localhost:9092",
+		Topic:             "",
+		DeliveryGuarantee: AT_LEAST_ONCE,
+		SaslAuthType:      SASL_NONE,
+	}
+	if err := cast.MapToStruct(props, c); err != nil {
+		return err
+	}
+	if len(strings.Split(c.Brokers, ",")) == 0 {
+		return fmt.Errorf("brokers can not be empty")
+	}
+	if c.Topic == "" {
+		return fmt.Errorf("topic can not be empty")
+	}
+	if !(c.DeliveryGuarantee == AT_MOST_ONCE || c.DeliveryGuarantee == AT_LEAST_ONCE) {
+		return fmt.Errorf("deliveryGuarantee incorrect")
+	}
+	if !(c.SaslAuthType == SASL_NONE || c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) {
+		return fmt.Errorf("saslAuthType incorrect")
+	}
+	if (c.SaslAuthType == SASL_SCRAM || c.SaslAuthType == SASL_PLAIN) && (c.SaslUserName == "" || c.SaslPassword == "") {
+		return fmt.Errorf("username and password can not be empty")
+	}
+
+	m.c = c
+	return nil
+}
+
+func (m *kafkaSink) Open(ctx api.StreamContext) error {
+	ctx.GetLogger().Debug("Opening kafka sink")
+	var maxAttempts int
+	var err error
+	var mechanism sasl.Mechanism
+	//delivery guarantee
+	switch m.c.DeliveryGuarantee {
+	case AT_MOST_ONCE:
+		maxAttempts = 1
+	case AT_LEAST_ONCE:
+		maxAttempts = 10
+	default:
+		maxAttempts = 10
+	}
```

Comment:
Is it ok to use `maxAttempts` to guarantee the delivery? Or we should expose `maxAttempts` instead of `DeliveryGuarantee`

Reply:
i mock the configuration of apache flink kafka connector , i think its ok to keep it ? https://nightlies.apache.org/flink/flink-docs-master/zh/docs/connectors/datastream/kafka/#kafka-sink

Reply:
maybe i should just delete this guarantee param and only offer a best effort delivery